dev/dram/device.c

```diff
@@ -93,75 +112,429 @@ void dram_cbt_exit(dram_dev_t *dram,
+
+    // Generate CA data packets to send training sequence
+    time_offset = (dram->cfg->t_vref_ca_long - dram->cfg->t_sh_train) >> ratio;
+    offset_remainder = ((dram->cfg->t_vref_ca_long - dram->cfg->t_sh_train) & (cycles_per_packet - 1));
+    // Add one extra DFI cycle if tDHtrain is odd value
+    if (offset_remainder > 0)
+    {
+        time_offset++;
+    }
+
+    packet = create_ck_packet_sequence(buffer, time_offset);
+    if (packet == NULL)
+    {
+        return WDDR_ERROR;
+    }
+
+    // Command
+    packet->packet.packet.cs_p0 = CS_HIGH << cs;
+    packet->packet.packet.cs_p1 = CS_HIGH << cs;
+    packet->packet.packet.address_p0 = command_address;
+
+    create_ck_packet_sequence(buffer, 1);
+
+    create_ck_packet_sequence(buffer, 2);
+
+    return WDDR_SUCCESS;
+}
+
+wddr_return_t dram_prepare_wrfifo_sequence(dram_dev_t *dram,
+                                           dfi_tx_packet_buffer_t *buffer,
+                                           burst_length_t burst_length,
+                                           chipselect_t cs,
+                                           uint16_t wrdata_en_offset,
+                                           command_data_t *data)
+{
+    uint16_t initial_dram_offset = 0;
+    uint8_t time_offset = 1;
+    packet_item_t *packet;
+    command_t command = {0};
+    packet_group_info_t wrdata_en_info, wrdata_info;
+
+    // Calculate timestamp of first CA packet (time 0).
+    // Shifted by ratio to be measured in DRAM CLK cycles.
+    initial_dram_offset = (buffer->ts_last_packet + time_offset) << dram->cfg->ratio;
+
+    // Calculate wr latency required for wr packet from time 0
+    // time 0 is identified as first CA packet ts
+    // wrdata offset is measured in DRAM CLKs
+    uint8_t wrdata_en_dram_clk = initial_dram_offset + CA_CLK_LEN + dram->cfg->phy_wr_lat + wrdata_en_offset;
+    uint8_t wrdata_dram_clk = wrdata_en_dram_clk + dram->cfg->phy_wr_en;
+    uint8_t cycles_per_packet = 1 << dram->cfg->ratio;
+
+    create_packet_group_info(&wrdata_en_info, dram->cfg->ratio, wrdata_en_dram_clk, burst_length + 4);
+    create_packet_group_info(&wrdata_info, dram->cfg->ratio, wrdata_dram_clk, burst_length);
+
+    // Create WRFIFO command sequence.
+    create_wrfifo_command(&command, data, cs);
+    create_address_packet_sequence(buffer, dram->cfg->ratio, &command, time_offset);
+
+    time_offset = wrdata_en_info.dfi_ts_start - buffer->ts_last_packet;
+
+    // If enough delay is required, insert CKE packet to keep clocks cycling.
+    if (time_offset > 1)
+    {
+        create_cke_packet_sequence(buffer, 1);
+        time_offset--;
+    }
+
+    while(wrdata_en_info.phase_remaining || wrdata_info.phase_remaining)
+    {
+        packet = create_cke_packet_sequence(buffer, time_offset);
+        if (packet == NULL)
+        {
+            return WDDR_ERROR;
+        }
+
+        if (buffer->ts_last_packet >= wrdata_en_info.dfi_ts_start)
```

Comment:
I think using timestamp in the packet looks better than accessing the buffer.

Reply:
agreed, I think this is a cleaner representation of what is happening but achieves the same result.  Updated to use packet time instead of last packet from buffer.